src/java.base/share/classes/java/util/LinkedHashMap.java

```diff
@@ -788,5 +788,21 @@ public final void remove() {
         public final Map.Entry<K,V> next() { return nextNode(); }
     }
 
+    /**
+     * Creates a new, empty LinkedHashMap suitable for the expected number of mappings.
+     * The returned map uses the default load factor of 0.75, and its initial capacity is
+     * generally large enough so that the expected number of mappings can be added
+     * without resizing the map.
+     *
+     * @param numMappings the expected number of mappings
+     * @param <K>         the type of keys maintained by this map
+     * @param <V>         the type of mapped values
+     * @return the newly created map
+     * @throws IllegalArgumentException if numMappings is negative
+     * @since 19
+     */
+    public static <K, V> LinkedHashMap<K, V> newLinkedHashMap(int numMappings) {
```

Comment:
`LinkedHashMap` may be often extended for it has a `protected boolean removeEldestEntry(Entry)`. Should we make a separate factory method for such instances (with functional implementation) or just expose `HashMap.calculateHashMapCapacity`?

Reply:
Good question. Having to subclass and override this method in order to provide a removal policy has always seemed rather clumsy. However, it's the supported approach, and it's done fairly frequently in the wild. A new subclass requires that the caller invoke `new` on that specific subclass, which in turn must choose which superclass constructor to call. This means that a static factory method can't be used. The alternatives would be to expose another constructor or to expose `calculateHashMapCapacity` as you suggest. A new constructor might also need to control the load factor and the ordering policy (insertion vs access order) so that's a fairly complicated overload to consider.

Exposing the calculate method might help but that's mostly just a wrapper around a small computation. As we've seen it's easy to get this computation wrong, but exposing a method that _just_ does this computation seems like a really narrow case.

(Still another alternative would be to pass a lambda expression that's called at the appropriate time. That would involve adding a `BiPredicate<Map<K,V>, Map.Entry<K,V>>` to yet another constructor overload. This could work but it doesn't seem any simpler.)

The need for subclassing LinkedHashMap and overriding this method might also be reduced by the addition of new APIs from the Sequenced Collections proposal (https://openjdk.java.net/jeps/8280836). One simply needs to call `pollFirstEntry` at the right time. That might remove the need to have some expiration policy plugged directly into the map itself.

I'm not inclined to add more APIs to cover what seems to be a fairly narrow case, but we might keep this in mind to see if anything useful pops up.

Reply:
True. My initial idea was to pass a lambda expression if we do make another factory.

However, when I look at the constructor parameters, I realized that for those evicting-queue-like use cases, it's more preferable for users to just call

```java
new LinkedHashMap<>(cacheSize + 1, 1, true) {
    protected boolean removeEldestEntry(Entry<K, V> entry) {
        return size() > cacheSize;
    }
}
```
which preallocates the cache, avoids growth when the cache is full, and removes by least-recently-used.

IMO this use case is better suited by the existing APIs (like the call above) than a new factory method, especially that many libraries, like guava or caffeine, offer more efficient caches.

So yes, this patch in the current state looks good. In the JDK, it seems there are a few cases of linked hash map evicting by size that can have the map initialized to be presized and avoid growth. But that's off-topic here.